assets/aml-benchmark/components/src/dataset_preprocessor/dataset_preprocessor.py

```diff
@@ -0,0 +1,136 @@
+parent_folder = os.path.dirname(current_folder)
+sys.path.append(parent_folder)
+logger = get_logger(__name__)
+
+jinja2.filters.FILTERS['zip'] = zip
+ENV = jinja2.Environment()
+ENV.globals.update(zip=zip)
+
+
+class DatasetPreprocessor(object):
+    """DatasetPrerprocessor object class."""
+
+    def __init__(
+        self,
+        input_dataset: str = None,
+        template: str = None,
+        user_preprocessor: str = None,
+        encoder_config: str = None,
+        output_dataset: str = None
+    ):
+        """Dataset Preprocessor Class.
+
+        :param input_dataset: Path to the directory to load the dataset.
+        :param template: A json dictionary where key is the name of the column enclosed in " " and associated \
+            dict value is presented using jinja template logic which will be used to extract the \
+            respective value from the dataset.
+        :param user_preprocessor: Path to the custom preprocessor python script provided by user.
+        :param encoder_config: Dictionary in json format that contains idtolabel or labeltoid to encode. This is not \
+            applicable to custom scripts.
+        :param output_dataset: Path to the dump the processed .jsonl file.
+        """
+        self.input_dataset = input_dataset
+        self.template = template
+        self.user_preprocessor = user_preprocessor
+        self.encoder_config = encoder_config
+        self.output_dataset = output_dataset
+        self.validate()
+
+    def validate(self):
+        """Validate the parameters."""
+        if self.input_dataset is None:
+            mssg = (
+                "Path to load the dataset is not provided."
+            )
+            raise BenchmarkValidationException._with_error(
+                AzureMLError.create(BenchmarkValidationError, error_details=mssg)
+            )
+        if self.template is None and self.user_preprocessor is None:
+            mssg = (
+                "Please provide the input to apply preprocessing logic either via template input or \
+                script_path."
+            )
+            raise BenchmarkValidationException._with_error(
+                AzureMLError.create(BenchmarkValidationError, error_details=mssg)
+            )
+        if self.user_preprocessor and not self.user_preprocessor.endswith('.py'):
+            mssg = (
+                "Please provide python script containing your custom preprocessor logic."
+            )
+            raise BenchmarkValidationException._with_error(
+                AzureMLError.create(BenchmarkValidationError, error_details=mssg)
+            )
+
+    def add_json_filter(self, template):
+        """Add tojson filter in the template."""
+        patt = r'({{.*?}})'
+        pat = re.compile(patt)
+        matches = re.findall(pat, template)
+        for m in range(len(matches)):
+            logger.info(f"{template}, {matches[m]}")
+            s1 = '{{'+matches[m].lstrip('{{').rstrip('}}')+'|tojson}}'

```

Comment:
There is a lot of quite subtle stuff going on, like regex matching, string replacements, etc. Would be good to cover with unit tests.

Reply:
Added.